L1Trigger/TrackerTFP/src/MiniHoughTransform.cc

```diff
@@ -0,0 +1,285 @@
+          it = stream.erase(remove(next(stream.begin(), pos), it, nullptr), it);
+        it = next(it, s);
+      }
+      // adjust stream start so that first output stub is in first place in case of quickest track
+      if (!stream.empty())
+        stream.erase(stream.begin(), next(stream.begin(), setup_->mhtMinLayers()));
+    }
+  }
+
+  // mux 4 streams to 1 stream
+  void MiniHoughTransform::slb(vector<deque<StubMHT*>>& inputs, vector<StubMHT*>& accepted, TTDTC::Stream& lost) const {
+    if (all_of(inputs.begin(), inputs.end(), [](const deque<StubMHT*>& stubs){ return stubs.empty(); }))
+      return;
+    auto size = [](int& sum, const deque<StubMHT*>& stubs){ return sum += stubs.size(); };
+    const int nFrames = accumulate(inputs.begin(), inputs.end(), 0, size);
+    accepted.reserve(nFrames);
+    // input fifos
+    vector<deque<StubMHT*>> stacks(numCells_);
+    // helper for handshake
+    TTBV empty(-1, numCells_, true);
+    TTBV enable(0, numCells_);
+    // clock accurate firmware emulation, each while trip describes one clock tick, one stub in and one stub out per tick
+    while(!all_of(inputs.begin(), inputs.end(), [](const deque<StubMHT*>& d){ return d.empty(); }) or
+          !all_of(stacks.begin(), stacks.end(), [](const deque<StubMHT*>& d){ return d.empty(); })) {
+      // store stub in fifo
+      for(int channel = 0; channel < numCells_; channel++){
+        StubMHT* stub = pop_front(inputs[channel]);
+        if (stub)
+          stacks[channel].push_back(stub);
+      }
+      // identify empty fifos
+      for (int channel = 0; channel < numCells_; channel++)
+        empty[channel] = stacks[channel].empty();
+      // chose new fifo to read from if current fifo got empty
+      const int iEnableOld = enable.plEncode();
+      if (enable.none() || empty[iEnableOld]) {
+        enable.reset();
+        const int iNotEmpty = empty.plEncode(false);
+        if (iNotEmpty < numCells_)
+          enable.set(iNotEmpty);
+      }
+      // read from chosen fifo
+      const int iEnable = enable.plEncode();
+      if (enable.any())
+        accepted.push_back(pop_front(stacks[iEnable]));
+      else
+        // gap if no fifo has been chosen
+        accepted.push_back(nullptr);
+
+    }
+    // perform truncation if desired
+    if (enableTruncation_ && (int)accepted.size() > setup_->numFrames()) {
+      const auto limit = next(accepted.begin(), setup_->numFrames());
+      auto valid = [](int& sum, StubMHT* stub){ return sum += stub ? 1 : 0; };
+      const int nLost = accumulate(limit, accepted.end(), 0, valid);
+      lost.reserve(nLost);
+      for (auto it = limit; it != accepted.end(); it++)
+        if (*it)
+          lost.emplace_back((*it)->frame());
+      accepted.erase(limit, accepted.end());
+    }
+    // cosmetics -- remove gaps at the end of stream
+    for(auto it = accepted.end(); it != accepted.begin();)
+      it = (*--it) == nullptr ? accepted.erase(it) : accepted.begin();
+  }
+
+  //
```

Comment:
Add comment "Dynamic load balancing of inputs". And in 1 sentence indicate what this balancing is doing physically.

Reply:
done